tests/components/onvif/test_device_trigger.py

```diff
@@ -0,0 +1,124 @@
+"""The tests for ONVIF device triggers."""
+import pytest
+
+import homeassistant.components.automation as automation
+from homeassistant.components.onvif import DOMAIN
+from homeassistant.components.onvif.device import ONVIFDevice
+from homeassistant.components.onvif.event import (
+    CONF_ONVIF_EVENT,
+    CONF_UNIQUE_ID,
+    Event,
+    EventManager,
+)
+from homeassistant.helpers import device_registry
+from homeassistant.setup import async_setup_component
+
+from tests.common import (
+    MockConfigEntry,
+    assert_lists_same,
+    async_get_device_automations,
+    async_mock_service,
+    mock_device_registry,
+)
+
+
+@pytest.fixture
+def device_reg(hass):
+    """Return an empty, loaded, registry."""
+    return mock_device_registry(hass)
+
+
+@pytest.fixture
+def calls(hass):
+    """Track calls to a mock service."""
+    return async_mock_service(hass, "test", "automation")
+
+
+async def test_get_triggers(hass, device_reg):
+    """Test we get the expected triggers from a onvif."""
+    # Add ONVIF config entry
+    onvif_config_entry = MockConfigEntry(
+        domain=DOMAIN, unique_id="12:34:56:AB:CD:EF", data={}
+    )
+    onvif_config_entry.add_to_hass(hass)
+    device_entry = device_reg.async_get_or_create(
+        config_entry_id=onvif_config_entry.entry_id,
+        connections={(device_registry.CONNECTION_NETWORK_MAC, "12:34:56:AB:CD:EF")},
+    )
+
+    # Add another config entry for this device
+    unifi_config_entry = MockConfigEntry(
+        domain="unifi", unique_id="12:34:56:AB:CD:EF", data={}
+    )
+    unifi_config_entry.add_to_hass(hass)
+    device_reg.async_get_or_create(
+        config_entry_id=unifi_config_entry.entry_id,
+        connections={(device_registry.CONNECTION_NETWORK_MAC, "12:34:56:AB:CD:EF")},
+    )
+
+    # Create device and mock event
+    device = ONVIFDevice(hass, onvif_config_entry)
+    device.events = EventManager(hass, None, onvif_config_entry.unique_id)
+    # pylint: disable=protected-access
+    device.events._events = {
+        f"{onvif_config_entry.unique_id}_tns1:RuleEngine/LineDetector/Crossed_0_0_0": Event(
+            f"{onvif_config_entry.unique_id}_tns1:RuleEngine/LineDetector/Crossed_0_0_0",
+            "0 Line Crossed",
+            "event",
+        )
+    }
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][onvif_config_entry.unique_id] = device
```

Comment:
It's better to set up the integration so we don't need to mock integration code.

Reply:
The events would need to be mocked anyway as they come from the device.  I could call integration setup here and mock the `ONVIFDevice.async_setup` method since that is where the communication happens.

Reply:
We prefer patching in the periphery ie the library classes and methods. Then we don't need to change tests if we refactor the integration.